### ISSUE TITLE
Undesirable garbage collector/spawnProcess interaction

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2483]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2483|
|Type|defect|
|Created|2007-02-25 21:13:47Z|

Finalizers can run in the child process created by spawnProcess:

```
from os import getpid

from twisted.internet import reactor
from twisted.internet.utils import getProcessOutputAndValue
from twisted.internet.task import LoopingCall

pid = getpid()
print pid

class BadObject(object):
    def __del__(self):
        if getpid() != pid:
            print getpid(), '!=', pid


class CycleParticipant(object):
    pass


def makeSomeGarbage():
    a = CycleParticipant()
    b = CycleParticipant()
    a.b = b
    b.a = a
    a.bad = BadObject()


def runChildProcess():
    makeSomeGarbage()
    d = getProcessOutputAndValue("/bin/false")
    def cbOutput((out, err, code)):
        if out or err:
            print 'Unexpected:', out, err
    d.addCallback(cbOutput)
    return d


def main():
    d = LoopingCall(runChildProcess)
    d.start(0).addErrback(lambda ign: reactor.stop())
    reactor.run()

if __name__ == '__main__':
    main()
```

That they can run there at all is bad and probably surprising to most people.  As a specific wrinkle, they can run either before the file descriptors have been hooked up to the parent, in which case the output will just go to the parent's output (resulting in double output, since the finalizer will presumably run and do the same thing in the parent), or afterwards, in which case the parent will actually /see/ output notionally (but not actually) generated by itself as coming from the child, or it can happen while the file descriptors are being set up, in which case some random unpredictable thing will happen.

<details><summary>Searchable metadata</summary>

```
trac-id__2483 2483
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1172438027000000 1172438027000000
changetime__1194633023000000 1194633023000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

I came across this because of a failure on one of the buildslaves.  twisted.test.test_process.ProcessTestCase.testManyProcesses failed because it saw the output from numerous temporary file objects being garbage collected (and failing) in the output from its child process.

This poorly written test reported this in the log file:

```
twisted.test.test_process.ProcessTestCase.testManyProcesses ... err != 1234: "Exception exceptions.IOError: (9, 'Bad file descriptor') in <bound method _TemporaryFileWrapper.__del__ of <closed file '<fdopen>', mode 'w+b' at 0xad51dbf0>> ignored\nException exceptions.IOError: (9, 'Bad file descriptor') in <bound method _TemporaryFileWrapper.__del__ of <closed file '<fdopen>', mode 'w+b' at 0xad51dba8>> ignored\nException exceptions.IOError: (9, 'Bad file descriptor') in <bound method _TemporaryFileWrapper.__del__ of <closed file '<fdopen>', mode 'w+b' at 0xad51dc80>> ignored\nException exceptions.IOError: (9, 'Bad file descriptor') in <bound method _TemporaryFileWrapper.__del__ of <closed file '<fdopen>', mode 'w+b' at 0xad51dd10>> ignored\nException exceptions.IOError: (9, 'Bad file descriptor') in <bound method _TemporaryFileWrapper.__del__ of <closed file '<fdopen>', mode 'w+b' at 0xad51dd58>> ignored\n1234"
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_2"></a>|@jyknight commented|
|-|-|

So, gc.disable() before forking?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

That's the only thing that's come to mind so far.  I guess it's not really very bad.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

I added the necessary gc calls, and tests.  Code is in **spawnprocess-gc-2483**.  I feel kind of bad about this code, but I'm not totally sure why.  Probably because the spawnProcess code is so hairy, complex, undertested, and _still_ buggy.  After this is merged I'll fix some of the other bugs nearby, like leaked pipes and incorrect set{u,g}id calls.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

test_mockWithForkErrorGarbageCollectorDisabled docstring is talking about the 'enable' case, it should be corrected.

Otherwise, everything looks fine, except I don't see a real test for the actual problem.

I'm relatively convinced it fixes the reported behavior, but having a deterministic test to show it would be great. I think it's probably possible with a forced gc.collect() somewhere (maybe simply mocking _setupChild or _execChild ?).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [21683]) docstring cleanups

refs [#2483](https://github.com/twisted/twisted/issues/2483)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun removed owner|
|-|-|

Fixed the docstring.  Re-reading it, I found some other bits of the wording confusing, so I changed slightly more than the enable/disable mistake.

I still can't think of any better gc test to write.  As discussed on IRC, aside from checking to make sure it is disable at the right times, I can't think of any way to test that the garbage collector isn't invoked in the forked child.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set owner to @exarkun|
|-|-|

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [21686]) Merge spawnprocess-gc-2483

Author: exarkun
Reviewer: therve
Fixes [#2483](https://github.com/twisted/twisted/issues/2483)

Disable garbage collection completely in the forked child used to implement
IReactorProcess.spawnProcess on POSIX to prevent finalizers or weakref
callbacks from executing in that process.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
